notebooks/03_evaluate/comparison_movielens.ipynb

```diff
@@ -0,0 +1,982 @@
+    "                                       relevancy_method=\"top_k\")\n",
+    "    return {\n",
+    "        \"MAP\": rank_eval.map_at_k(),\n",
+    "        \"nDCG@k\": rank_eval.ndcg_at_k(),\n",
+    "        \"Precision@k\": rank_eval.precision_at_k(),\n",
+    "        \"Recall@k\": rank_eval.recall_at_k()\n",
+    "    }\n",
+    "    \n",
+    "    \n",
+    "def rating_metrics_python(test, predictions):\n",
+    "    cols = {\n",
+    "        \"col_user\": USER_COL, \n",
+    "        \"col_item\": ITEM_COL, \n",
+    "        \"col_rating\": RATING_COL, \n",
+    "        \"col_prediction\": PREDICTION_COL\n",
+    "    }\n",
+    "    return {\n",
+    "        \"RMSE\": rmse(test, predictions, **cols),\n",
+    "        \"MAE\": mae(test, predictions, **cols),\n",
+    "        \"R2\": rsquared(test, predictions, **cols),\n",
+    "        \"Explained Variance\": exp_var(test, predictions, **cols)\n",
+    "    }\n",
+    "    \n",
+    "    \n",
+    "def ranking_metrics_python(test, predictions, k=10):\n",
+    "    cols = {\n",
+    "        \"col_user\": USER_COL, \n",
+    "        \"col_item\": ITEM_COL, \n",
+    "        \"col_rating\": RATING_COL, \n",
+    "        \"col_prediction\": PREDICTION_COL\n",
+    "    }\n",
+    "    return {\n",
+    "        \"MAP\": map_at_k(test, predictions, k=k, **cols),\n",
+    "        \"nDCG@k\": ndcg_at_k(test, predictions, k=k, **cols),\n",
+    "        \"Precision@k\": precision_at_k(test, predictions, k=k, **cols),\n",
+    "        \"Recall@k\": recall_at_k(test, predictions, k=k, **cols)\n",
+    "    }\n",
+    "    \n",
+    "    \n",
+    "rating_evaluator = {\n",
+    "    \"als\": lambda test, predictions: rating_metrics_pyspark(test, predictions),\n",
+    "    \"svd\": lambda test, predictions: rating_metrics_python(test, predictions),\n",
+    "    \"fastai\": lambda test, predictions: rating_metrics_python(test, predictions)\n",
+    "}\n",
+    "    \n",
+    "    \n",
+    "ranking_evaluator = {\n",
+    "    \"als\": lambda test, predictions, k: ranking_metrics_pyspark(test, predictions, k),\n",
+    "    \"sar_single_node\": lambda test, predictions, k: ranking_metrics_python(test, predictions, k),\n",
+    "    \"svd\": lambda test, predictions, k: ranking_metrics_python(test, predictions, k),\n",
+    "    \"fastai\": lambda test, predictions, k: ranking_metrics_python(test, predictions, k),\n",
+    "    \"ncf\": lambda test, predictions, k: ranking_metrics_python(test, predictions, k),\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def generate_summary(data, algo, k, train_time, time_rating=np.nan, rating_metrics=None, time_ranking=np.nan, ranking_metrics=None):\n",
+    "    summary = {\"Data\": data, \"Algo\": algo, \"K\": k, \"Train time\": train_time, \"Rating time\": time_rating, \"Ranking time\": time_ranking}\n",
+    "    if rating_metrics is None:\n",
```

Comment:
Again here you repeat the names of the metrics. What about defining two lists of rating and ranking metrics in the beginning of this notebook and using dict comprehensions etc. in all other places? If you want to add another metric in the future, you will need to scan through the notebook and find all occurrences of the metrics.